candidate-agreement/src/table.rs

```diff
@@ -0,0 +1,986 @@
+	}
+
+	/// Get an iterator of all candidates with a given group.
+	// TODO: impl iterator
+	pub fn candidates_in_group<'a>(&'a self, group_id: C::GroupId)
+		-> Box<Iterator<Item=&'a CandidateData<C>> + 'a>
+	{
+		Box::new(self.candidate_votes.values().filter(move |c| c.group_id == group_id))
+	}
+
+	/// Drain all misbehavior observed up to this point.
+	pub fn drain_misbehavior(&mut self) -> HashMap<C::ValidatorId, <C as ResolveMisbehavior>::Misbehavior> {
+		::std::mem::replace(&mut self.detected_misbehavior, HashMap::new())
+	}
+
+	/// Import a signed statement. Signatures should be checked for validity, and the
+	/// sender should be checked to actually be a validator.
+	///
+	/// This can note the origin of the statement to indicate that he has
+	/// seen it already.
+	pub fn import_statement(
+		&mut self,
+		context: &C,
+		statement: SignedStatement<C::Candidate, C::Digest, C::ValidatorId, C::Signature>,
+		from: Option<C::ValidatorId>
+	) -> Option<Summary<C::Digest, C::GroupId>> {
+		let SignedStatement { statement, signature, sender: signer } = statement;
+
+		let trace = match statement {
+			Statement::Candidate(_) => StatementTrace::Candidate(signer.clone()),
+			Statement::Valid(ref d) => StatementTrace::Valid(signer.clone(), d.clone()),
+			Statement::Invalid(ref d) => StatementTrace::Invalid(signer.clone(), d.clone()),
+			Statement::Available(ref d) => StatementTrace::Available(signer.clone(), d.clone()),
+		};
+
+		let (maybe_misbehavior, maybe_summary) = match statement {
+			Statement::Candidate(candidate) => self.import_candidate(
+				context,
+				signer.clone(),
+				candidate,
+				signature
+			),
+			Statement::Valid(digest) => self.validity_vote(
+				context,
+				signer.clone(),
+				digest,
+				ValidityVote::Valid(signature),
+			),
+			Statement::Invalid(digest) => self.validity_vote(
+				context,
+				signer.clone(),
+				digest,
+				ValidityVote::Invalid(signature),
+			),
+			Statement::Available(digest) => self.availability_vote(
+				context,
+				signer.clone(),
+				digest,
+				signature,
+			)
+		};
+
+		if let Some(misbehavior) = maybe_misbehavior {
+			// all misbehavior in agreement is provable and actively malicious.
+			// punishments are not cumulative.
+			self.detected_misbehavior.insert(signer, misbehavior);
+		} else {
+			if let Some(from) = from {
+				self.note_trace_seen(trace.clone(), from);
+			}
+
+			self.note_trace_seen(trace, signer);
+		}
+
+		maybe_summary
+	}
+
+	fn note_trace_seen(&mut self, trace: StatementTrace<C::ValidatorId, C::Digest>, known_by: C::ValidatorId) {
+		self.validator_data.entry(known_by).or_insert_with(|| ValidatorData {
+			proposal: None,
+			known_statements: HashSet::default(),
+		}).known_statements.insert(trace);
+	}
+
+	fn import_candidate(
+		&mut self,
+		context: &C,
+		from: C::ValidatorId,
+		candidate: C::Candidate,
+		signature: C::Signature,
+	) -> (Option<<C as ResolveMisbehavior>::Misbehavior>, Option<Summary<C::Digest, C::GroupId>>) {
```

Comment:
consider defining the return type as a struct since it appears to be returned from multiple functions.